internal/ingest/segmented_sync.go

```diff
@@ -0,0 +1,194 @@
+package ingest
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-legs"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	"github.com/ipld/go-ipld-prime/traversal/selector/builder"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var (
+	ssb        = builder.NewSelectorSpecBuilder(basicnode.Prototype.Any)
+	adSequence = ssb.ExploreFields(
+		func(efsb builder.ExploreFieldsSpecBuilder) {
+			efsb.Insert("PreviousID", ssb.ExploreRecursiveEdge())
+		})
+	entriesSequence = ssb.ExploreFields(func(efsb builder.ExploreFieldsSpecBuilder) {
+		efsb.Insert("Next", ssb.ExploreRecursiveEdge()) // Next field in EntryChunk
+	})
+)
+
+func (ing *Ingester) syncEntries(ctx context.Context, pid peer.ID, c cid.Cid, peerAddr multiaddr.Multiaddr, hookFunc legs.BlockHookFunc) error {
+
+	// If segmented sync is disabled, sync the entries in one go according to the configured
+	// entries depth limit.
+	nextDepthLimit := ing.cfg.SyncSegmentDepthLimit
+	if nextDepthLimit == -1 {
+		log.Debug("segmented sync is disabled; syncing entries in one call")
+		entSel := ssb.ExploreRecursive(recursionLimit(ing.cfg.EntriesDepthLimit), entriesSequence).Node()
+		_, err := ing.sub.Sync(ctx, pid, c, entSel, peerAddr, legs.ScopedBlockHook(hookFunc))
+		return err
+	}
+
+	// If the configured total max depth is not unlimited and is less than next segment depth limit
+	// then use the configured total depth limit to avoid overshooting sync depth.
+	totalDepthLimit := ing.cfg.EntriesDepthLimit
+	if totalDepthLimit != -1 && totalDepthLimit < ing.cfg.SyncSegmentDepthLimit {
+		nextDepthLimit = totalDepthLimit
+	}
+
+	// Use blockHook to learn a synced entry chunk's "Next" link.
+	var nextChunkLink *ipld.Link
+	var latestLoadErr error
+	hookOpt := legs.ScopedBlockHook(func(id peer.ID, c cid.Cid) {
+		// Load the chunk to learn what its next chunk link is.
+		// Load cannot be done later because entries are removed as soon as they are processed, and
+		// they are processed as they are synced.
+		var chunk *schema.EntryChunk
+		chunk, latestLoadErr = ing.loadEntryChunk(c)
+		if latestLoadErr == nil {
+			nextChunkLink = chunk.Next
+		}
+		if hookFunc != nil {
+			hookFunc(id, c)
+		}
+	})
+
+	nextSegmentCid := c
+	var depthSoFar int
+	for {
+		sel := ssb.ExploreRecursive(selector.RecursionLimitDepth(int64(nextDepthLimit)), entriesSequence).Node()
+		_, err := ing.sub.Sync(ctx, pid, nextSegmentCid, sel, peerAddr, hookOpt)
+		if err != nil {
+			return err
+		}
+		depthSoFar += nextDepthLimit
+
+		// If the latest chunk load failed in block hook then return it here.
+		if latestLoadErr != nil {
+			log.Error("failed to load chunk during segmented sync", "err", err)
```

Comment:
does this mean on failure we aren't caching the subset we have already and start over from the beginning of the sync? is there a way to have failures still cache the subset that was gotten and resume from there?

Reply:
The blocks are written into datastore right up to the point where a potential failure happens. We can implement some resumption strategy here where:
* given a CID, we recursively walk its chain (either `Next` for `EntryChunk` or `PreviousID` for `Advertisement`)
* for each link we check if it is present in the datastore
* start segmented sync from the first link that is not found in the datastore.

A potential issues I see is that background gorutines might have already processed and removed those ads from the datastore.

Thoughts?

Reply:
happy with this being in a subsequent follow-up PR.

eventually with storing blocks in e.g. a local IPFS data store i'd hope the blocks wouldn't get removed once they're there.

Reply:
That'd be great; that'd also make the sync more efficient in that we don't need to decode every cid during sync to learn its next node in the chain since processing might remove it if we defer the load

Reply:
https://github.com/filecoin-project/storetheindex/issues/492